e2e/app/element/base-element.ts

```diff
@@ -115,59 +123,68 @@ export default class BaseElement {
+      const boxModel = await element.boxModel();
+      return boxModel;
     };
 
     return this.asElementHandle()
-      .then((element) => {
-        return isDisplayed(element) && !!boxModel(element);
+      .then(async (element) => {
+        return (await isDisplayed(element)) && (await boxModel(element)) !== null;
       })
       .catch(() => false);
   }
 
   async click(options?: ClickOptions): Promise<void> {
-    return this.asElementHandle().then(async (element) => {
-      // Click workaround: Wait for x,y to stop changing within specified time
+    await this.asElementHandle().then(async (element) => {
+      // Click workaround: Wait for (x,y) to stop changing within specified time
       const startTime = Date.now();
       let previousX: number;
       let previousY: number;
       while (Date.now() - startTime < 30 * 1000) {
-        const viewport = await element.isIntersectingViewport();
-        if (viewport) {
-          const boundingBox = await element.boundingBox();
-          const boxModel = await element.boxModel();
-          if (boundingBox && boxModel) {
-            const x = boundingBox.x + boundingBox.width / 2;
-            const y = boundingBox.y + boundingBox.height / 2;
-            if (previousX !== undefined && previousY !== undefined) {
-              if (
-                parseFloat(previousX.toFixed(7)) === parseFloat(x.toFixed(7)) &&
-                parseFloat(previousY.toFixed(7)) === parseFloat(y.toFixed(7))
-              ) {
-                break;
-              }
+        const visible = await this.isVisible();
+        if (!visible) {
+          // Scrolls element into viewport if needed.
+          await element.hover();
+          await element.focus();
+          await this.page.waitForTimeout(1000);
+        }
+        const boundingBox = await element.boundingBox();
+        const boxModel = await element.boxModel();
+        if (boundingBox !== null && boxModel !== null) {
+          const x = boundingBox.x + boundingBox.width / 2;
+          const y = boundingBox.y + boundingBox.height / 2;
+          if (previousX !== undefined && previousY !== undefined) {
+            if (
+              parseFloat(previousX.toFixed(7)) === parseFloat(x.toFixed(7)) &&
+              parseFloat(previousY.toFixed(7)) === parseFloat(y.toFixed(7))
+            ) {
+              break;
             }
-            previousX = x;
-            previousY = y;
           }
+          previousX = x;
+          previousY = y;
         }
-        await element.hover();
+        // Half second sleep before getting values of (x, y) again.
         await this.page.waitForTimeout(500);
       }
-      return element.click(options);
+      // End the test if previousX or previousY is still undefined after waiting.
+      if (previousX === undefined || previousY === undefined) {
+        throw new Error(`Failed to click element because it is not visible. xpath: ${this.getXpath()}`);
+      }
```

Comment:
I didn't realize until now the values of previousX and previousY could still be `undefined` after the while loop has reached max wait time. Clicking on a non-visible element just doesn't work. Puppeteer doesn't throw any error from click. It makes test flaky.